src/bin/languageserver/mod.rs

```diff
@@ -1158,26 +1168,52 @@ impl LanguageServer for SolangServer {
     async fn did_open(&self, params: DidOpenTextDocumentParams) {
         let uri = params.text_document.uri;
 
+        if let Ok(path) = uri.to_file_path() {
+            self.files
+                .lock()
+                .await
+                .text_buffers
+                .insert(path, params.text_document.text);
+        }
+
         self.parse_file(uri).await;
     }
 
     async fn did_change(&self, params: DidChangeTextDocumentParams) {
         let uri = params.text_document.uri;
 
+        if let Ok(path) = uri.to_file_path() {
+            if let Some(text_buf) = self.files.lock().await.text_buffers.get_mut(&path) {
+                *text_buf = params
+                    .content_changes
+                    .into_iter()
+                    .fold(text_buf.clone(), update_file_contents);
```

Comment:
Usually the `.clone()` here is unnecessary. `fn update_file_contents()` could take `&str` rather than `String`, and this would save having to clone the text every time.

Reply:
`fold` requires the function that is passed (in this case, `update_file_contents`) to have the same same input and output type. So, I had to use `clone` on `text_buf` to make both input and output to `update_file_contents` a `String`. https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold
I decided to use `fold` as I believe it makes the code more readable.